Drivers/Src/Display.c

```diff
@@ -1,69 +1,190 @@
- * @param buf the output buffer
- * @return void
+static const char *DELIMITER = ".";
+static const char *ASSIGNMENT = "=";
+static const char *TERMINATOR = "\xff\xff\xff";
+static const char *NO_ERROR = "No Error";
+
+// Color defintions for display
+static const uint16_t NEXTION_GREEN = 2016;
+static const uint16_t NEXTION_RED = 63488;
+static const uint16_t NEXTION_LIGHT_GREY = 42260;
+static const uint16_t NEXTION_DARK_GREY = 23275;
+static const uint16_t NEXTION_BURNT_ORANGE = 51872;
+
+static inline int IsNextionFailure(uint32_t val) {
+    return ((val & ~0x00FFFFFF) != (1 << 24));
+}
+
+
+// List of names of possible strings to build a command out of
+enum CommandString_t {
+    VELOCITY,
+    CRUISE_ENABLE,
+    CRUISE_SET,
+    REGEN_ENABLE,
+    VALUE,
+    TEXT,
+    PCO,
+    SYSTEM,
+    PAGE,
+    ERROR0,
+    ERROR1,
+    ERROR2,
+    ERROR3,
+    ERROR4,
+    ERROR5
+};
+
+// The command strings themselves
+static char *CommandStrings[] = {
+    "x0",
+    "t1",
+    "t2",
+    "t3",
+    "val",
+    "txt",
+    "pco",
+    "",
+    "page",
+    "t4",
+    "t5",
+    "t6",
+    "t7",
+    "t8",
+    "t9"
+};
+
+/**
+ * Convert signed 32-bit integer to string
+ * Returns pointer to string
+ */
+static char *to_charp(int32_t num, char *dest) {
+    bool neg = false;
+    if (num < 0) {
+        neg = true;
+        num = -num;
+    }
+    int index = 12;
+    dest[--index] = '\0';
+    do {
+        dest[--index] = (num % 10) + '0';
+        num /= 10;
+    } while (num > 9);
+    if (num != 0) dest[--index] = num + '0';
+    if (neg) dest[--index] = '-';
+    return dest + index; // return pointer to beginning of string
+}
+
+/**
+ * Sets an object's attribute to a value
  */
-static void encodeByte(uint8_t byte, uint8_t *buf) {
-    uint8_t hi = (byte >> 4) & 0x0F;
-    uint8_t lo = byte & 0x0F;
-    buf[0] = hi < 10 ? (hi + '0') : (hi - 10 + 'A');
-    buf[1] = lo < 10 ? (lo + '0') : (lo - 10 + 'A');
-    // e.g encoding the byte `42` will write the characters
-    // '2' and 'A' to the locations pointed at by buf and buf+1
+static ErrorStatus updateValue(enum CommandString_t obj_index, enum CommandString_t attr_index, char *msg, int32_t val) {
+    char buf[12]; // To store converted int
+    char *number = to_charp(val, buf); // Convert integer to string
+    char *obj = CommandStrings[obj_index];
+    char *attr = CommandStrings[attr_index];
+    int len1 = strlen(obj);
+    int len2 = strlen(attr);
+
+    // If not modifying a global, send obj
+    if (len1 != 0) {
+        BSP_UART_Write(UART_3, obj, len1);
+        BSP_UART_Write(UART_3, (char *) DELIMITER, 1);
+    }
+
+    BSP_UART_Write(UART_3, attr, len2); // Send the attribute
+    BSP_UART_Write(UART_3, (char *) ASSIGNMENT, 1);
+    // If msg is valid, send it; otherwise send a converted number
+    if (msg != NULL) {
+        BSP_UART_Write(UART_3, msg, strlen(msg));
+    } else {
+        BSP_UART_Write(UART_3, number, strlen(number)); // Send the value
+    }
```

Comment:
Would it make sense to make a few versions of this function (eg. `updateObj()`, `updateMessage()`, `updateValue()`), so you can provide a nicer interface and don't have to do these checks to figure out what you are trying to update?